### PR TITLE
1.5 PR

### PR DESCRIPTION
**Pull request by prdpsvs**
Updated dbt-tests-adapter version to 1.5.2
Updated dbt_version to 1.5 and dbt core dependency to 1.5.2
Updated adapter version to 1.5.0-rc1
Added constraint support to FabricAdapter class
Added get_column_schema_from_query to test data type mapping.
Remove get_column_schema_from_query post testing from FabricAdapter.
Added data type mapping between pyodbc & SQL.
Implemented data_type_code_to_name in Connection Manager class.
Added 3 methods to columns.sql to assert model and constraints in yaml file.
Added contact config check in create_table_As.
use default__assert_columns_equivalent().
Added test_constraints to test the contract.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,8 +1,19 @@
 # Changelog
 
+### v1.5.0-rc1
+
+* Upgraded dbt-fabric adapter to match dbt-core & dbt-tests-adapter version 1.5.2.
+* Added constraint support to dbt-fabric adapter.
+    * Check constraints are not supported.
+    * Column & model constraints are not supported in CREATE TABLE command by Microsoft Fabric Data Warehouse. Column and model constraints are implemented by ALTER TABLE ADD Constraints command.
+    * SP_RENAME is not available at the time of this release. Until SP_RENAME is supported, the name of the constraints is determined by the adapter. Upon SP_RENAME support, the adapter will adhere to constraints name, if provided by user in model/column constraints.
```
Comment (dataders) [inline CHANGELOG.md:9]: language tweak

```suggestion
    * user-defined names for constraints are not currently supported. naming is handled by the adapter, until `SP_RENAME` is supported in Fabric
```

File: dbt/adapters/fabric/fabric_adapter.py [hunk 2]
```diff
@@ -18,9 +24,44 @@ class FabricAdapter(SQLAdapter):
...
+
+    @classmethod
+    def convert_number_type(cls, agate_table, col_idx):
+        decimals = agate_table.aggregate(agate.MaxPrecision(col_idx))
+        return "float" if decimals else "int"
+
     def create_schema(self, relation: BaseRelation) -> None:
         relation = relation.without_identifier()
-        fire_event(SchemaCreation(relation=_make_ref_key_msg(relation)))
+        # fire_event(SchemaCreation(relation=_make_ref_key_msg(relation)))
```
Comment (dataders) [inline dbt/adapters/fabric/fabric_adapter.py:64]: why is this commented out?

File: dbt/adapters/fabric/fabric_connection_manager.py [hunk 3]
```diff
@@ -30,6 +30,27 @@
...
+    # "int": "smallint",
+    # "int": "tinyint",
+    "int": "int",
+    "bytes": "varbinary",
+    "bool": "bit",
+    "datetime.date": "date",
+    "datetime.datetime": "datetime2(6)",
+    "datetime.time": "time",
+    "decimal.Decimal": "decimal",
+    # "decimal.Decimal": "numeric",
```
Comment (dataders) [inline dbt/adapters/fabric/fabric_connection_manager.py:51]: I'm assuming the commenting is from your testing? that said, if there's any knowledge in your brain about why specific types are used opposed to others, here would be a great place to add that info as a comment for... posterity

```suggestion
    "str": "varchar",
    "uuid.UUID": "uniqueidentifier",
    "uuid": "uniqueidentifier",
    "float": "bigint",
    "int": "int",
    "bytes": "varbinary",
    "bool": "bit",
    "datetime.date": "date",
    "datetime.datetime": "datetime2(6)",
    "datetime.time": "time",
    "decimal.Decimal": "decimal",
```

File: dbt/include/fabric/macros/adapters/columns.sql [hunk 4]
```diff
@@ -1,3 +1,15 @@
...
+    {% if sql.strip().lower().startswith('with') %}
+        {{ select_sql }}
+    {% else -%}
+        select * from (
+        {{ select_sql }}
+    ) dbt_sbq_tmp
+    where 1 = 0
+    {%- endif -%}
+
+{% endmacro %}
```
Comment (dataders) [inline dbt/include/fabric/macros/adapters/columns.sql:11]: this is very clever! can you add a docstring or comment for why this is needed?

File: dbt/include/fabric/macros/adapters/relation.sql [hunk 5]
```diff
@@ -83,6 +80,74 @@
...
+
+      {%call statement('drop_table_constraints', fetch_result= True)%}
+        SELECT drop_constraint_statement FROM
+        (
+          SELECT 'ALTER TABLE ['+TABLE_SCHEMA+'].['+TABLE_NAME+'] DROP CONSTRAINT ' + CONSTRAINT_NAME AS drop_constraint_statement
+          FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS
+          WHERE TABLE_NAME = '{{ from_relation.identifier }}' and TABLE_SCHEMA = '{{ from_relation.schema }}'
+        ) T WHERE drop_constraint_statement IS NOT NULL
+
+      {% endcall %}
```
Comment (dataders) [inline dbt/include/fabric/macros/adapters/relation.sql:124]: whoa -- this looks like it took a lot of time to work out! is this all necessary for renaming a relation?

**Review by dataders — COMMENTED** [review]
just had a few nits and questions, but this looks great. really impressed with what you've been able to ship. hopefully you have a clear prioritized of TSQL API features that would make these features easier

File: dbt/adapters/fabric/fabric_adapter.py (hunk 2 above)
Comment (prdpsvs) [reply dbt/adapters/fabric/fabric_adapter.py]: I addressed it in the latest commit. dbt core 1.5 does not support _make_ref_key_msg https://github.com/dbt-labs/dbt-core/blob/1.5.latest/core/dbt/adapters/cache.py#L5

File: dbt/adapters/fabric/fabric_connection_manager.py (hunk 3 above)
Comment (prdpsvs) [reply dbt/adapters/fabric/fabric_connection_manager.py:51]: Although the same key can have multiple values, the data_type_code_to_name picks the first value for asserting model constraints. I have aligned the tests to match uncommented data types but want to keep others for future reference.

File: dbt/include/fabric/macros/adapters/columns.sql (hunk 4 above)
Comment (prdpsvs) [reply dbt/include/fabric/macros/adapters/columns.sql:11]: Some of the models are expressed as a CTE. SELECT * FROM (<CTE>) is not supported by Microsoft sql engines. So, I added a check to see if the sub query is CTE or not.

File: dbt/include/fabric/macros/adapters/relation.sql (hunk 5 above)
Comment (prdpsvs) [reply dbt/include/fabric/macros/adapters/relation.sql:124]: SP_RENAME is not supported yet. I am using CTAS and DROP TABLE to manage SP_RENAME. However, the contraints are created on __dbt_tmp/__dbt_backup relations but on final models. Hence, I have to re-create constraints on the final table before dropping __dbt_tmp/__dbt_backup relations.

This is not necessary as soon as SP_RENAME is supported (expecting to be available by September).

**Comment by prdpsvs** [comment]
> just had a few nits and questions, but this looks great. really impressed with what you've been able to ship. hopefully you have a clear prioritized of TSQL API features that would make these features easier

Thanks Anders, I updated the PR based on your feedback. Could you please review?

**Pull request closed without merging**